airflow/contrib/operators/aws_lambda_operator.py

```diff
@@ -0,0 +1,131 @@
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+#
+
+from airflow.models import BaseOperator
+from airflow.utils.decorators import apply_defaults
+from airflow.contrib.hooks.aws_lambda_hook import AwsLambdaHook
+
+
+class AwsLambdaExecutionError(Exception):
+    """
+    Raised when there is an error executing the function.
+    """
+
+    pass
+
+
+class AwsLambdaPayloadError(Exception):
+    """
+    Raised when there is an error with the Payload object in the response.
+    """
+
+    pass
+
+
+class AwsLambdaInvokeFunctionOperator(BaseOperator):
+    """
+    Invoke AWS Lambda functions with a JSON payload.
+
+    The check_success_function signature should be a single param which will receive a dict.
+    The dict will be the "Response Structure" described in
+    https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/lambda.html#Lambda.Client.invoke.
+    It may be necessary to read the Payload see the actual response from the Lambda function e.g.,
+
+    ```
+    def succeeded(response):
+        payload = json.loads(response['Payload'].read())
+        # do something with payload
+    ```
+
+    :param function_name: The name of the Lambda function.
+    :type function_name: str
+    :param region_name: AWS region e.g., eu-west-1, ap-southeast-1, etc.
+    :type region_name: str
+    :param payload: The JSON to submit as input to a Lambda function.
+    :type payload: str
+    :param check_success_function: A function to check the Lambda response and determine success or failure.
+    :type check_success_function: function
+    :param log_type: Set to Tail to include the execution log in the response. Otherwise, set to "None".
+    :type log_type: str
+    :param qualifier: A version or alias name for the Lambda.
+    :type qualifier: str
+    :param aws_conn_id: connection id of AWS credentials / region name. If None,
+        credential boto3 strategy will be used
+        (http://boto3.readthedocs.io/en/latest/guide/configuration.html).
+    :type aws_conn_id: str
+    """
+
+    @apply_defaults
+    def __init__(
+        self,
+        function_name,
+        region_name,
+        payload,
+        check_success_function,
+        log_type="None",
+        qualifier="$LATEST",
+        aws_conn_id=None,
+        *args,
+        **kwargs,
+    ):
+        super().__init__(*args, **kwargs)
+        self.function_name = function_name
+        self.payload = payload
+        self.check_success_function = check_success_function
+        self.aws_lambda_hook = AwsLambdaHook(
```

Comment:
The initialization of the hook should be in the `execute` imo. The `__init__` should only save given arguments to the instance (`self`) to be available when executing the operation.

Reply:
What is the advantage of initialising it in the `execute` method instead of in the constructor? In other languages I would stay away from initialising dependencies in the constructor and use dependency injection instead. Originally I used a property to initialise the hook and call that from `execute` but I removed it in a round of refactoring because it was overkill. Since the hook is only ever used once inside `execute` it felt simpler to just initialise it in the constructor. It meant setting less properties, too. I could just declare the hook as a param in the constructor but that's just moving the problem around. Previous code looked something like below:

    @property
    def aws_lambda_hook(self):
        if not hasattr(self, "_aws_lambda_hook"):
            self._aws_lambda_hook = AwsLambdaHook(...)
        return self._aws_lambda_hook

Reply:
This causes this object to be created when it is not needed. 
This happens in two cases:
- when all DAG on the scheduler is refreshed. Sometimes it even happens every eg. 120 seconds.
- when each individual task from current DAG is started, it loads the entire DAG again.
It is therefore very important to optimize all processes so that the objects are the lightest. Recently, we also had a problem with some operators, because they initiated the hook, and the hook execute a database query.
https://issues.apache.org/jira/browse/AIRFLOW-4771
It is worth realizing that the `execute` method is executed only once, so do not add additional caches for objects.
Another thing is the division of roles between the hook and the operator. The operator is rarely inherited because it often calls only hook methods. Hook is very important because it provides an interface for the authors of their own operators.
Operators are created according to the idea: "Prefer composition over inheritance". 
https://en.wikipedia.org/wiki/Composition_over_inheritance